syntax = "proto3";

package panicker;

// –ë–∞–∑–æ–≤—ã–µ —Å–æ–æ–±—â–µ–Ω–∏—è
message Ticker {
    string symbol = 1;
    string name = 2;
}

message Candle {
    string ticker = 1;
    double open = 2;
    double high = 3;
    double low = 4;
    double close = 5;
    int64 volume = 6;
    string timestamp = 7;
    string interval = 8;
}

message PanicSignal {
    enum SignalType {
        PANIC = 0;
        GREED = 1;
        NEUTRAL = 2;
    }

    enum Level {
        STRONG = 0;
        MODERATE = 1;
        URGENT = 2;
        IGNORE = 3;
    }

    string ticker = 1;
    SignalType signal_type = 2;
    Level level = 3;
    double rsi_14 = 4;
    double rsi_7 = 5;
    double rsi_21 = 6;
    double volume_ratio = 7;
    double current_price = 8;
    string detected_at = 9;
    double risk_metric = 10;
    repeated VolumeCluster volume_clusters = 11;
    string interpretation = 12;
}

message VolumeCluster {
    double price_level = 1;
    double volume_percentage = 2;
    string role = 3; // "support", "resistance", "neutral"
}

message OverheatIndex {
    string ticker = 1;
    double overheat_percentage = 2;  // ‚Üê –∏—Å–ø—Ä–∞–≤–ª–µ–Ω–æ
    double current_rsi = 3;
    double volume_ratio = 4;
    string last_signal_time = 5;
    PanicSignal.Level last_signal_level = 6;
}

message ScanRequest {
    repeated Ticker tickers = 1;
    bool real_time = 2;
}

message ScanResponse {
    repeated PanicSignal signals = 1;
    string scan_id = 2;
    string timestamp = 3;
    int32 total_scanned = 4;
    int32 signals_found = 5;
}

// –°–µ—Ä–≤–∏—Å –¥–ª—è —Ä–∞–±–æ—Ç—ã —Å –ø–∞–Ω–∏–∫–æ–π
service PanickerService {
    // –û—Å–Ω–æ–≤–Ω–æ–π –º–µ—Ç–æ–¥ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è
    rpc ScanTickers(ScanRequest) returns (ScanResponse);

    // –ü–æ–ª—É—á–∏—Ç—å —Å—Ç–∞—Ç—É—Å –∑–¥–æ—Ä–æ–≤—å—è –∞–∫—Ü–∏–∏
    rpc GetOverheatIndex(Ticker) returns (OverheatIndex);

    // –ü–æ–ª—É—á–∏—Ç—å –∏—Å—Ç–æ—Ä–∏—é —Å–∏–≥–Ω–∞–ª–æ–≤
    rpc GetSignalHistory(HistoryRequest) returns (SignalHistory);
}

// –°–µ—Ä–≤–∏—Å –¥–ª—è —Ä—ã–Ω–æ—á–Ω—ã—Ö –¥–∞–Ω–Ω—ã—Ö
service MarketDataService {
    // –ü–æ–ª—É—á–∏—Ç—å —Å–≤–µ—á–∏ –¥–ª—è —Ç–∏–∫–µ—Ä–∞
    rpc GetCandles(CandleRequest) returns (CandleResponse);

    // –ü–æ–ª—É—á–∏—Ç—å —Ç–µ–∫—É—â–∏–µ —Ü–µ–Ω—ã
    rpc GetCurrentPrices(PriceRequest) returns (PriceResponse);

    // –ü–æ–ª—É—á–∏—Ç—å –¥–∞–Ω–Ω—ã–µ —Å—Ç–∞–∫–∞–Ω–∞
    rpc GetOrderBook(OrderBookRequest) returns (OrderBookResponse);
}

service SignalsService {
    // –ü–æ—Ç–æ–∫–æ–≤–∞—è –ø–µ—Ä–µ–¥–∞—á–∞ —Å–∏–≥–Ω–∞–ª–æ–≤ –≤ —Ä–µ–∞–ª—å–Ω–æ–º –≤—Ä–µ–º–µ–Ω–∏
    rpc StreamSignals(StreamRequest) returns (stream PanicSignal);

    // –ü–æ–ª—É—á–∏—Ç—å —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤
    rpc GetTopSignals(TopRequest) returns (TopResponse);

    // –ü–æ–ª—É—á–∏—Ç—å —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫—É –∑–∞ –ø–µ—Ä–∏–æ–¥
    rpc GetStats(StatsRequest) returns (StatsResponse);

    // –ò–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞—Ç—å —Ç–∏–∫–µ—Ä
    rpc IgnoreTicker(IgnoreRequest) returns (IgnoreResponse);
}

// –í—Å–ø–æ–º–æ–≥–∞—Ç–µ–ª—å–Ω—ã–µ —Å–æ–æ–±—â–µ–Ω–∏—è
message HistoryRequest {
    string ticker = 1;
    int32 days_back = 2;
    string start_date = 3;
    string end_date = 4;
    int32 limit = 5;
}

message SignalHistory {
    repeated PanicSignal signals = 1;
    int32 total_count = 2;
}

message CandleRequest {
    string ticker = 1;
    string interval = 2;
    int32 count = 3;
}

message CandleResponse {
    repeated Candle candles = 1;
    string request_id = 2;
}

message PriceRequest {
    repeated string tickers = 1;
}

message PriceResponse {
    map<string, double> prices = 1;
    string timestamp = 2;
}

message OrderBookRequest {
    string ticker = 1;
    int32 depth = 2;
}

message OrderBookResponse {
    string ticker = 1;
    repeated OrderBookEntry bids = 2;
    repeated OrderBookEntry asks = 3;
    double spread_percentage = 4;
}

message OrderBookEntry {
    double price = 1;
    int64 quantity = 2;
}

message StreamRequest {
    repeated string tickers = 1;
    bool include_all = 2;
}

message TopRequest {
    string period = 1; // "today", "week", "month"
    int32 limit = 2;
}

message TopResponse {
    repeated PanicSignal top_signals = 1;
    string period = 2;
}

message IgnoreRequest {
    string ticker = 1;
    int32 duration_hours = 2;
}

message IgnoreResponse {
    bool success = 1;
    string ignored_until = 2;
}
// ============================================================================
// –°–¢–ê–¢–ò–°–¢–ò–ö–ê
// ============================================================================

message StatsRequest {
    int32 days = 1;  // –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –¥–Ω–µ–π –¥–ª—è —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏
}

message StatsResponse {
    int32 total_signals = 1;          // –í—Å–µ–≥–æ —Å–∏–≥–Ω–∞–ª–æ–≤
    int32 strong_signals = 2;         // üî¥ –°–ò–õ–¨–ù–´–•
    int32 moderate_signals = 3;       // üü° –£–ú–ï–†–ï–ù–ù–´–•
    int32 urgent_signals = 4;         // ‚ö™ –°–†–û–ß–ù–´–•
    string most_active_ticker = 5;    // –°–∞–º—ã–π –∞–∫—Ç–∏–≤–Ω—ã–π —Ç–∏–∫–µ—Ä
    int32 most_active_count = 6;      // –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Å–∏–≥–Ω–∞–ª–æ–≤ —É —Å–∞–º–æ–≥–æ –∞–∫—Ç–∏–≤–Ω–æ–≥–æ
    string most_calm_ticker = 7;      // –°–∞–º—ã–π —Å–ø–æ–∫–æ–π–Ω—ã–π —Ç–∏–∫–µ—Ä (—Å —Å–∏–≥–Ω–∞–ª–∞–º–∏)
    int32 most_calm_count = 8;        // –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Å–∏–≥–Ω–∞–ª–æ–≤ —É —Å–∞–º–æ–≥–æ —Å–ø–æ–∫–æ–π–Ω–æ–≥–æ
    string market_tension = 9;        // –û–±—â–∞—è –Ω–∞–ø—Ä—è–∂—ë–Ω–Ω–æ—Å—Ç—å —Ä—ã–Ω–∫–∞
}